### ISSUE TITLE
Chasing an UB-smelling bug related to swap_rows

### ISSUE THREAD

**Issue by HadrienG2** [rustsim/nalgebra#396]
To the best of my knowledge, the following two code snippets (where `mat` is an owned mutable nalgebra Matrix) should produce the same final result:

```
// Clean version
mat.swap_rows(r1, r2);

// Ugly version
let row1 = mat.fixed_rows::<U1>(r1).into_owned();
let row2 = mat.fixed_rows::<U1>(r2).into_owned();
mat.fixed_rows_mut::<U1>(r1).copy_from(&row2);
mat.fixed_rows_mut::<U1>(r2).copy_from(&row1);
```

Unfortunately, I have a program where the former produces different (and wrong) results, with suspiciously UB-ish symptoms: if I add a print statement near the code that does the swap_rows, the results become correct.

I do not use unsafe myself, so I can only assume that the issue is in nalgebra. But looking at the source code of swap_rows did not reveal anything suspicious, and I was not successful in producing a simpler reproducer that I can share with you so far.

If you have more experience in debugging this kind of issues than I do, any tips on how to narrow down the problem further would be much appreciated.

**Comment by HadrienG2** [comment]
I did not manage to produce a simple reproducer yet, but the full application is on Github if you feel like playing around with it:

- Check out the `annotate_sort` branch of https://github.com/HadrienG2/3photons-rust/tree/annotate_sort .
- Uncomment [the print statement at line 192 of src/event.rs](https://github.com/HadrienG2/3photons-rust/blob/annotate_sort/src/event.rs#L192).
- Run the application in release mode. You will see what output the row-sorting code should produce.
- Now, comment out the print statement and run in release mode again. On my computer, this produces a different result.
- The `event-matrix` branch of the same repo applies the workaround above and runs correctly (you can rebase `annotate_sort` on top of it if you want to compare).

There is no question that sorting the rows of a column-major matrix is highly questionable from a performance point of view (it's a long story). However, it should not miscompile! :)

**Comment by sebcrozet** [comment]
Also note that preventing inlining (`#[inline(never)]`) of the matrix `swap_rows` method also prevents the bug from happening.

**Comment by HadrienG2** [comment]
One hypothesis which I have concerning to this bug, which your observations tend to fuel, is that it _might_ be related to the use of &mut.

Material on unsafe Rust says that &mut must not alias, but do not go much further than that. This leaves some dangerous margin of interpretation for the compiler. For example, in our case, can a load/store of a certain size be extended into a masked load/store of a larger size for optimization, even if it potentially creats some aliasing which did not exist before?

To be more precise, Rust references are currently defined by the language reference to [do what LLVM's noalias does](https://doc.rust-lang.org/reference/behavior-considered-undefined.html). But LLVM's noalias [is described in compiler-speak that is not very accessible to mere mortals](https://www.llvm.org/docs/LangRef.html#noalias-and-alias-scope-metadata) and its implementation [historically had many bugs which stopped rustc from using it until very recently](https://github.com/rust-lang/rust/issues/31681).

TL;DR: Instead of optimized assembly, which is highly sensitive to environmental conditions as we have both observed, maybe the right approach would be to take a simple swap_rows() example, generate LLVM IR from it, and go see the compiler team / unsafe code guidelines team to ask them if the IR really does what we think it does.

**Comment by HadrienG2** [comment]
Ah, finally, a minimal reproducer!

```
extern crate nalgebra;

use nalgebra::Matrix3x4;

#[inline(never)]
fn swappy() -> Matrix3x4<f32> {
    let mut mat = Matrix3x4::new(1., 2.,  3.,  4.,
                                 5., 6.,  7.,  8.,
                                 9., 10., 11., 12.);

    for i in 0..2 {
        for j in i..3 {
            if mat[(j, 3)] > mat[(i, 3)] { mat.swap_rows(i, j); }
        }
    }

    mat
}

fn main() {
    let mat = swappy();
    println!("Result: {}", mat);
}
```

**Comment by HadrienG2** [comment]
According to the discussion that occured on https://github.com/nikomatsakis/rust-memory-model/issues/50, the code follows Rust's aliasing rules, and this is therefore a compiler bug. So, to rust-lang/rust I go next...

**Comment by sebcrozet** [comment]
Thanks for all your investigations @HadrienG2 !

**Comment by HadrienG2** [comment]
Latest development at rust-lang/rust: The bug was reproduced with safe code, so this is _definitely_ a compiler bug.

**Comment by HadrienG2** [comment]
A workaround which disables assumption of no mutable aliasing ("noalias") for &mut until the root cause is found has landed in beta, should be part of the next stable release.

**Comment by pietroalbini** [comment]
This will also be fixed in Rust 1.29.2, which should be released tomorrow.

**Comment by HadrienG2** [comment]
1.29.2 has now been officially released, and I can confirm that it fixes this bug.

**Comment by sebcrozet** [comment]
Great ! We can close this then. Thanks for reporting and following this so closely.
